### PR TITLE
Prefiltering added to phylophlan_metagenomic.py

### PR DESCRIPTION
**Pull request by kmladenovic**

### CODE REVIEW

File: phylophlan/phylophlan_metagenomic.py [hunk 1]
```diff
@@ -741,6 +867,10 @@ def phylophlan_metagenomic():
 
         args.database = os.path.join(args.database_folder, args.database)
         args.mapping = os.path.join(args.database_folder, args.mapping)
+        args.db_pref = os.path.join(args.database_folder, 'dbs_pref')
```
Comment (fasnicar) [inline phylophlan/phylophlan_metagenomic.py:870]:
```suggestion
        args.db_pref = os.path.join(args.database_folder, args.database + '_pref')
```

File: phylophlan/phylophlan_metagenomic.py [hunk 2]
```diff
@@ -749,8 +879,11 @@ def phylophlan_metagenomic():
                                               nproc=args.nproc, verbose=args.verbose)
         args.database = args.output_prefix + '_sketches'
 
+
```
Comment (fasnicar) [inline phylophlan/phylophlan_metagenomic.py:882]:
```suggestion
```

File: phylophlan/phylophlan_metagenomic.py [hunk 3]
```diff
@@ -523,13 +527,132 @@ def disting_rec(x):
...
+                raise
+
+        with open('unassigned.txt', 'w') as f:
+            for mag in input_list:
+                f.write(f"{mag.strip('.msh')} is likely close to the SGB in full database, but SGB in not available in ChochoPhlAn\n")
+
+        if verbose:
+            t1 = time.time()
+            info('Inputs pasted in {}s\n'.format(int(t1 - t0)))
+
```
Comment (fasnicar) [inline phylophlan/phylophlan_metagenomic.py:598]:
```suggestion


```

File: phylophlan/phylophlan_metagenomic.py [hunk 4]
```diff
@@ -523,13 +527,132 @@ def disting_rec(x):
...
+        
+        table=table[table['dist'] < 0.2]
+        table['sgb'] = table['sgb'].map(lambda x: x.split("__")[0].strip('genomes/SGB'))
+        table['mag'] = table['mag'].map(lambda x: x.split('/')[-1].strip('.'+input_extension) + '.msh')
+
+        input_list=list(table['mag'].unique())
+        for sgb in chocophlan_list:
+            genomes = [output_prefix+'_sketches/inputs/'+ g for g in table['mag'][table['sgb']==sgb].values]
+            input_list = [x for x in input_list if output_prefix+'_sketches/inputs/' + x not in genomes]
+
```
Comment (fasnicar) [inline phylophlan/phylophlan_metagenomic.py:565]:
```suggestion
```

File: phylophlan/phylophlan_metagenomic.py [hunk 5]
```diff
@@ -523,13 +527,132 @@ def disting_rec(x):
...
+    for i in os.listdir(os.path.join(output_prefix + "_prefiltering/pref_dbs")):
+        table = pd.read_csv(os.path.join(output_prefix + "_prefiltering/pref_dbs"+'/'+i), sep='\t',names=['sgb','mag','dist','pvalue','hits'])
+        
+        table=table[table['dist'] < 0.2]
+        table['sgb'] = table['sgb'].map(lambda x: x.split("__")[0].strip('genomes/SGB'))
+        table['mag'] = table['mag'].map(lambda x: x.split('/')[-1].strip('.'+input_extension) + '.msh')
+
+        input_list=list(table['mag'].unique())
+        for sgb in chocophlan_list:
```
Comment (fasnicar) [inline phylophlan/phylophlan_metagenomic.py:562]:
```suggestion

        for sgb in chocophlan_list:
```

File: phylophlan/phylophlan_metagenomic.py [hunk 6]
```diff
@@ -523,13 +527,132 @@ def disting_rec(x):
...
+    outf = output_prefix + "_prefiltering/" + "{}_paste"
+    inpf = output_prefix + "_prefiltering/{}_genomes_inputs_list.txt"
+
+    for i in os.listdir(os.path.join(output_prefix + "_prefiltering/pref_dbs")):
+        table = pd.read_csv(os.path.join(output_prefix + "_prefiltering/pref_dbs"+'/'+i), sep='\t',names=['sgb','mag','dist','pvalue','hits'])
+        
+        table=table[table['dist'] < 0.2]
+        table['sgb'] = table['sgb'].map(lambda x: x.split("__")[0].strip('genomes/SGB'))
+        table['mag'] = table['mag'].map(lambda x: x.split('/')[-1].strip('.'+input_extension) + '.msh')
+
```
Comment (fasnicar) [inline phylophlan/phylophlan_metagenomic.py:560]:
```suggestion
```

File: phylophlan/phylophlan_metagenomic.py [hunk 7]
```diff
@@ -523,13 +527,132 @@ def disting_rec(x):
...
+        terminating.set()
+
+# pasting inputs per sgb
+def prefiltering_pasting(output_prefix, input_extension, chocophlan_list, verbose=True): 
+    outf = output_prefix + "_prefiltering/" + "{}_paste"
+    inpf = output_prefix + "_prefiltering/{}_genomes_inputs_list.txt"
+
+    for i in os.listdir(os.path.join(output_prefix + "_prefiltering/pref_dbs")):
+        table = pd.read_csv(os.path.join(output_prefix + "_prefiltering/pref_dbs"+'/'+i), sep='\t',names=['sgb','mag','dist','pvalue','hits'])
+
```
Comment (fasnicar) [inline phylophlan/phylophlan_metagenomic.py:556]:
```suggestion
```

File: phylophlan/phylophlan_metagenomic.py [hunk 8]
```diff
@@ -476,42 +482,40 @@ def pasting(output_prefix, prj_name, verbose=False):
...
-def disting_rec(x):
+        info('Prefiltering mash dist already computed!\n')
+    
+def prefiltering_rec(x):
     if not terminating.is_set():
         try:
-            pasted_bins, sgb_msh_idx, dist_file, verbose = x
-
+            msh_idx, db_pref, dist_file, verbose = x
+
```
Comment (fasnicar) [inline phylophlan/phylophlan_metagenomic.py:509]:
```suggestion

```

File: phylophlan/phylophlan_metagenomic.py [hunk 9]
```diff
@@ -476,42 +482,40 @@ def pasting(output_prefix, prj_name, verbose=False):
...
+        info('Prefiltering mash dist already computed!\n')
+    
+def prefiltering_rec(x):
     if not terminating.is_set():
         try:
-            pasted_bins, sgb_msh_idx, dist_file, verbose = x
-
+            msh_idx, db_pref, dist_file, verbose = x
+            
             if not os.path.isfile(dist_file):
```
Comment (fasnicar) [inline phylophlan/phylophlan_metagenomic.py:510]:
```suggestion

            if not os.path.isfile(dist_file):
```

File: phylophlan/phylophlan_metagenomic.py [hunk 10]
```diff
@@ -476,42 +482,40 @@ def pasting(output_prefix, prj_name, verbose=False):
...
                 error(str(e), init_new_line=True)
-                error('disting crashed', init_new_line=True, exit=True)
+                error('prefiltering crashed', init_new_line=False, exit=False)
     else:
-        info('Mash dist already computed!\n')
-
-
-def disting_rec(x):
+        info('Prefiltering mash dist already computed!\n')
+
```
Comment (fasnicar) [inline phylophlan/phylophlan_metagenomic.py:504]:
```suggestion


```

File: phylophlan/phylophlan_metagenomic.py [hunk 11]
```diff
@@ -454,10 +458,11 @@ def pasting(output_prefix, prj_name, verbose=False):
     for inpc in glob.iglob(output_prefix + "_sketches/{}_inputs_list_*.txt".format(prj_name)):
         chunk = int(inpc[::-1].split('.')[1].split('_')[0][::-1])
         outc = outf.format(chunk)
-
+
```
Comment (fasnicar) [inline phylophlan/phylophlan_metagenomic.py:461]:
```suggestion

```

File: phylophlan/phylophlan_metagenomic.py [hunk 12]
```diff
@@ -454,10 +458,11 @@ def pasting(output_prefix, prj_name, verbose=False):
     for inpc in glob.iglob(output_prefix + "_sketches/{}_inputs_list_*.txt".format(prj_name)):
         chunk = int(inpc[::-1].split('.')[1].split('_')[0][::-1])
         outc = outf.format(chunk)
```
Comment (fasnicar) [inline phylophlan/phylophlan_metagenomic.py:460]:
```suggestion
        outc = outf.format(chunk)

```

File: phylophlan/phylophlan_metagenomic.py [hunk 13]
```diff
@@ -149,6 +151,7 @@ def check_params(args, verbose=False):
 
         if not args.mapping.endswith('.txt.bz2'):
             args.mapping += '.txt.bz2'
+
```
Comment (fasnicar) [inline phylophlan/phylophlan_metagenomic.py:154]:
```suggestion
```

File: phylophlan/phylophlan_metagenomic.py (hunk 1 above)
Comment (fasnicar) [reply phylophlan/phylophlan_metagenomic.py:870]: I think we can call the prefiltering db starting with the db name. Although, now it is unclear to me what folder we create in line 204. I think we can assume the structure of the tar archive we'll distribute. What do you think?

Comment (kaca-git) [reply phylophlan/phylophlan_metagenomic.py:870]: I agree about the name, this was just for the working version. I agree about the structure of the tar archive too. The folder created in 204 is where the mash dist between inputs and prefiltering dbs is saved, but we can change that.

**Pull request merged**